mypy/types.py

```diff
@@ -1507,3 +1534,38 @@ def true_or_false(t: Type) -> Type:
     new_t.can_be_true = type(new_t).can_be_true
     new_t.can_be_false = type(new_t).can_be_false
     return new_t
+
+
+def function_type(func: mypy.nodes.FuncBase, fallback: Instance) -> FunctionLike:
+    if func.type:
+        assert isinstance(func.type, FunctionLike)
+        return func.type
+    else:
+        # Implicit type signature with dynamic types.
+        # Overloaded functions always have a signature, so func must be an ordinary function.
+        assert isinstance(func, mypy.nodes.FuncItem), str(func)
+        fdef = cast(mypy.nodes.FuncItem, func)
+        name = func.name()
+        if name:
+            name = '"{}"'.format(name)
+
+        return CallableType(
+            [AnyType()] * len(fdef.arg_names),
+            fdef.arg_kinds,
+            fdef.arg_names,
+            AnyType(),
+            fallback,
+            name,
+            implicit=True,
+        )
+
+
+class TypeVarInstantiator(TypeTranslator):
```

Comment:
The partial types appear only when I use reveal_type, and it seems to be orthogonal to self-type specifically (it happens with normal generic functions too).